File: backend/app/schemas/contract_analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.contract_analysis import (
    ContractAnalysisResultStatus,
    ContractAnalysisRunStatus,
)


class ContractAnalysisEvidenceReferenceRead(
    BaseModel
):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    ordinal: int
    block_id: str
    block_ordinal: int
    start_character: int
    end_character: int
    quote: str
    quote_sha256: str


class ContractAnalysisFindingRead(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    finding_id: str
    ordinal: int
    category: str
    severity_level: str
    title: str
    description: str
    content_sha256: str
    evidence_references: list[
        ContractAnalysisEvidenceReferenceRead
    ]


class ContractAnalysisRunSummaryRead(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    contract_id: int
    document_version_id: int
    version_number: int
    created_by_user_id: int | None
    status: ContractAnalysisRunStatus
    executor: str
    model: str
    policy_id: str
    policy_version: str
    policy_sha256: str
    source_file_sha256: str | None
    extracted_text_sha256: str | None
    result_id: str | None
    result_status: (
        ContractAnalysisResultStatus | None
    )
    requires_human_review: bool | None
    content_sha256: str | None
    error_code: str | None
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None


class ContractAnalysisRunRead(
    ContractAnalysisRunSummaryRead
):
    findings: list[
        ContractAnalysisFindingRead
    ]

File: frontend/src/types/contractAnalysis.ts
export type ContractAnalysisRunStatus =
  | "running"
  | "completed"
  | "failed";

export type ContractAnalysisResultStatus =
  "machine_draft";

export type ContractAnalysisSeverity =
  | "low"
  | "medium"
  | "high"
  | "critical";

export interface ContractAnalysisEvidenceReference {
  id: number;
  ordinal: number;
  block_id: string;
  block_ordinal: number;
  start_character: number;
  end_character: number;
  quote: string;
  quote_sha256: string;
}

export interface ContractAnalysisFinding {
  id: number;
  finding_id: string;
  ordinal: number;
  category: string;
  severity_level: ContractAnalysisSeverity;
  title: string;
  description: string;
  content_sha256: string;
  evidence_references: ContractAnalysisEvidenceReference[];
}

export interface ContractAnalysisRunSummary {
  id: number;
  contract_id: number;
  document_version_id: number;
  version_number: number;
  created_by_user_id: number | null;
  status: ContractAnalysisRunStatus;
  executor: string;
  model: string;
  policy_id: string;
  policy_version: string;
  policy_sha256: string;
  source_file_sha256: string | null;
  extracted_text_sha256: string | null;
  result_id: string | null;
  result_status: ContractAnalysisResultStatus | null;
  requires_human_review: boolean | null;
  content_sha256: string | null;
  error_code: string | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface ContractAnalysisRun
  extends ContractAnalysisRunSummary {
  findings: ContractAnalysisFinding[];
}
